### PR TITLE
Perf event writes metric

### PR DESCRIPTION
**Pull request by geyslan**
### 1. Explain what the PR does

bae153278 **chore(performance): update tracee dashboard**
5a5c2e7ce **chore: add perfbuf metric per event (METRICS=1)**
5fe36d177 **chore(go.mod): bump api to latest 1f796cb**

5a5c2e7ce **chore: add perfbuf metric per event (METRICS=1)**

```
Enabled only when built with METRICS=1.

BPFPerfEventSubmitAttemptsCount and BPFPerfEventSubmitFailuresCount
count the number of events processed by the eBPF programs and written to
or attempted to be written to the perf buffer.

It is incremented right after the attempt of writing the event to the
perf buffer, making it possible to measure if the that event was
successfully written to the perf buffer or not.

This metric can be used to monitor the performance of individual eBPF
events and to detect potential bottlenecks.
```

image

### 2. Explain how to test it

`METRICS=1 make tracee`

`sudo ./dist/tracee -e socket_dup,execve --metrics --pyroscope --pprof`

`make -f builder/Makefile.performance dashboard-start`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -179,3 +179,5 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	kernel.org/pub/linux/libs/security/libcap/psx v1.2.70 // indirect
 )
+
+replace github.com/aquasecurity/tracee/api => ./api
```
Comment (geyslan) [inline go.mod:183]: This commit must be changed to reflect the new api.

File: api/v1beta1/diagnostic.proto [hunk 2]
```diff
@@ -12,11 +12,12 @@ message GetMetricsResponse {
 	uint64 EventsFiltered = 2;
 	uint64 NetCapCount = 3;
 	uint64 BPFLogsCount = 4;
-	uint64 ErrorCount = 5;
-	uint64 LostEvCount = 6; 
-	uint64 LostWrCount = 7; 
-	uint64 LostNtCapCount = 8; 
-	uint64 LostBPFLogsCount = 9;
+	uint64 BPFPerfEventWrites = 5;
```
Comment (geyslan) [inline api/v1beta1/diagnostic.proto:15]: This commit must be merged first in a separate PR.

File: Makefile [hunk 3]
```diff
@@ -501,6 +504,7 @@ $(OUTPUT_DIR)/tracee: \
 		-ldflags="$(GO_DEBUG_FLAG) \
 			-extldflags \"$(CGO_EXT_LDFLAGS_EBPF)\" \
 			-X github.com/aquasecurity/tracee/pkg/version.version=$(VERSION) \
+			-X github.com/aquasecurity/tracee/pkg/version.debug=$(DEBUG) \
```
Comment (geyslan) [inline Makefile:507]: Perhaps to change that flag to enable the metric with standard build. I mean, making possible to get that metric but with stripped binaries.

**Comment by geyslan** [comment]
After offline discussion, registering another way of approaching this: by enabling the metric with a flag instead of requiring a rebuild.

It's feasible considering an increment of program sizes, an additional branch check (metric enabled) and a dangling map.

**Comment by geyslan** [comment]
I'm working on a stress tool on top of this branch. Gonna let you know soon.

File: Makefile (hunk 3 above)
Comment (geyslan) [reply Makefile:507]: Update grafana dashboard.

File: pkg/metrics/collector.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/aquasecurity/tracee/pkg/counter"
+	"github.com/aquasecurity/tracee/pkg/logger"
+)
+
+type Collector struct {
+	m           sync.RWMutex
+	description string
+	values      map[uint64]uint64
```
Comment (geyslan) [inline pkg/metrics/collector.go:16]: @NDStrahilevitz I have fixed this key/value, but perhaps they can be generic. WDYT?

Comment (geyslan) [reply pkg/metrics/collector.go]: Already made it generic.

File: performance/dashboard/docker-compose.yml [hunk 5]
```diff
@@ -11,7 +9,7 @@ services:
     volumes:
       - /proc:/host/proc:ro
       - /sys:/host/sys:ro
-      - /:/rootfs:ro
+      - /:/rootfs:ro,rslave
```
Comment (geyslan) [inline performance/dashboard/docker-compose.yml:12]: Opportunistic change to accommodate new version:

`Error response from daemon: invalid mount config: must use either propagation mode "rslave" or "rshared" when mount source is within the daemon root, daemon root: "/var/lib/docker", bind mount source: "/", propagation: "rprivate"`

**Comment by geyslan** [comment]
@NDStrahilevitz @yanivagman I would appreciate your +1.

**Review by yanivagman — APPROVED** [review]
LGTM, see one nit comment

File: pkg/ebpf/tracee.go [hunk 6]
```diff
@@ -63,7 +64,9 @@ type Tracee struct {
 	running   atomic.Bool
 	done      chan struct{} // signal to safely stop end-stage processing
 	OutDir    *os.File      // use utils.XXX functions to create or write to this file
-	stats     metrics.Stats
+	// Metrics
+	stats *metrics.Stats
+	// Engine
```
[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/ebpf/tracee.go:69]: You're right. It's a leftover.
